rdrf/rdrf/testing/behaviour/features/auth/steps.py

```diff
@@ -0,0 +1,181 @@
+import logging
+import re
+import time
+from collections import OrderedDict
+
+from aloe import step, world
+from aloe.tools import guess_types
+from nose.tools import assert_equal, assert_true
+from selenium.webdriver.common.keys import Keys
+
+from rdrf.models.definition.models import Registry
+from rdrf.testing.behaviour.features.auth import utils
+from rdrf.testing.behaviour.features.auth.page import LoginPage, RequestPasswordResetPage, MailOutboxPage, \
+    ResetPasswordPage, BasePage, TwoFactorTokenGeneratorPage, TwoFactorLoginTokenPage, DisableTwoFactorAuthPage, \
+    get_site_links
+from rdrf.testing.behaviour.features.steps import click_link
+from rdrf.testing.behaviour.features.utils import scroll_to_y
+from registry.groups.models import CustomUser
+
+logger = logging.getLogger(__name__)
+
+
+@step('expecting to receive email')
+def setup_email(_step):
+    go_to_url(_step, 'mailbox_empty')
+    assert_true(MailOutboxPage(world.browser).is_displayed())
+
+
+@step('the registry is closed to registrations')
+def close_registration(_step):
+    utils.close_registration(world.registry)
+
+
+@step('the registry is open to registrations')
+def open_registration(_step):
+    utils.open_registration(world.registry)
+
+
+@step('go to the "([^"]+)" page')
+def go_to_url(_step, page_name):
+    page_url = get_site_links(world.site_url, world.registry_code).get(page_name)
+    world.browser.get(page_url)
+
+
+@step('try to register as a user called "([^"]+)" using the email address "([^"]+)" and the password "([^"]+)"')
+def patient_self_registration(_step, client_name, email_address, password):
+    go_to_url(_step, 'registration')
+
+    client_first_name = client_name.split()[0]
+    client_last_name = client_name.split()[1]
+
+    world.user_first_name = client_first_name
+    world.user_last_name = str.upper(client_last_name)
+
+    # Plain text field parameters
+    params = OrderedDict([
+        ('id_username', email_address),
+        ('id_password1', password),
+        ('id_password2', password),
+        ('id_first_name', client_first_name),
+        ('id_surname', client_last_name),
+        ('id_date_of_birth', '1980-09-01'),
+    ])
+
+    # Populate plain text fields
+    for key, value in params.items():
+        world.browser.find_element_by_id(key).send_keys(value + Keys.TAB)
+
+    # Select the gender radio button
+    # 1 - Male, 2 - Female, 3 - Indeterminate
+    world.browser.find_element_by_css_selector("input[name='gender'][value='1']").click()
+
+    captcha_iframe_element = world.browser.find_element_by_xpath(
+        "//iframe[@role='presentation']")
+
+    world.browser.switch_to.frame(captcha_iframe_element)
+    scroll_to_y(500)
+
+    world.browser.find_element_by_id('recaptcha-anchor').send_keys(Keys.SPACE)
+
+    time.sleep(4)
```

Comment:
I think it'd be better to wait for the `registration-submit` button to be enabled by `reCaptchaCallback`. Sleeping for a set amount of time can cause the test to fail when a related change alters the timing but there hasn't actually been a regression.

Reply:
This was existing (unused) code that I moved over, but I can have a look.